File: OJ/leet208/leet208v2.cpp
/*
leet208

Trie

*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <utility>
// #include "../utils/LeetCpp.utils.hpp"

using namespace std;


class Trie {
public:
    bool isleaf = false;
    Trie* next[26] {};

    /** Initialize your data structure here. */
    Trie() {}
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        auto p = this;
        for (const auto& ch: word) {
            if (p->next[ch-'a'] == nullptr) 
                p->next[ch-'a'] = new Trie();
            p = p->next[ch-'a'];
        }
        p->isleaf = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        auto p = this;
        for (const auto& ch: word) {
            if (p->next[ch-'a'] == nullptr) 
                return false;
            p = p->next[ch-'a'];
        }
        return p->isleaf;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        auto p = this;
        for (const auto& ch: prefix) {
            if (p->next[ch-'a'] == nullptr) 
                return false;
            p = p->next[ch-'a'];
        }
        return true;
    }
};


int main(int argc, char const *argv[])
{
    cout << boolalpha;

    Trie* tree = new Trie();
    tree->insert("apple");
    cout << tree->search("apple") << endl;
    cout << tree->search("app") << endl;
    cout << tree->startsWith("app") << endl;

    tree->insert("app");
    cout << tree->search("app") << endl;

    return 0;
}

File: OJ/leet208/leet208v2.py
# leet208

from typing import List


from collections import defaultdict
from functools import reduce

END = 0
class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        Trie = lambda : defaultdict(Trie)
        self.trie = Trie()


    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        reduce(dict.__getitem__, word, self.trie)[END] = word

    def __find(self, word: str) :
        
        p = self.trie
        for ch in word:
            if ch not in p: return None
            p = p[ch]
        return p  

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        return (p := self.__find(word)) is not None and END in p


    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        return self.__find(prefix) is not None


trie = Trie()

print(trie.insert("apple"))
print(trie.search("apple"))
print(trie.search("app"))
print(trie.startsWith("app"))
print(trie.insert("app"))
print(trie.search("app"))



